### PR TITLE
migration to stage-2 L1 for muon modules

### PR DESCRIPTION
**Pull request by sarafiorendi**
Muon specific modules using l1extra collections have been cloned and migrated to new stage-2 L1t collections.

HLTMuonL1Filter → HLTMuonL1TFilter
HLTL1MuonSelector → HLTL1TMuonSelector
HLTMuonL1RegionalFilter → HLTMuonL1TRegionalFilter
HLTMuonTrkFilter → HLTMuonTrkL1TFilter
HLTMuonL1toL3TkPreFilter → HLTMuonL1TtoL3TkPreFilter

HLTMuonL2ToL1Map → HLTMuonL2ToL1TMap
HLTMuonL2PreFilter → HLTMuonL2FromL1TPreFilter
HLTMuonDimuonL2Filter → HLTMuonDimuonL2FromL1TFilter

RecoMuon/L2MuonSeedGenerator/L2MuonSeedGenerator → L2MuonSeedGeneratorFromL1T

An update was also needed for 
DataFormats/MuonSeed/interface/L3MuonTrajectorySeed.h
DataFormats/MuonSeed/interface/L2MuonTrajectorySeed.h

### CODE REVIEW

**Comment by slava77** [comment]
Is the implementation really so different to justify cloned implementation?
It may be better to have updates to existing modules.

On the other hand, this is all just the trigger code.
So, continuing support and bugfixing to the past is going to be limited in time while comparisons are needed between the two and the newer one happens to need an update like a bug fix.

@battibas @jhgoh @HuguesBrun please confirm that this cloning decision was well discussed in muon POG.

File: DataFormats/MuonSeed/src/classes_def.xml [hunk 1]
```diff
@@ -27,7 +28,9 @@
       <class name="std::map<unsigned int,edm::helpers::KeyVal<edm::Ref<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed,edm::refhelper::FindUsingAdvance<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed> >,edm::RefVector<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed,edm::refhelper::FindUsingAdvance<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed> > > >"/>
       <class name="edm::helpers::KeyVal<edm::Ref<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed,edm::refhelper::FindUsingAdvance<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed> >,edm::RefVector<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed,edm::refhelper::FindUsingAdvance<std::vector<L2MuonTrajectorySeed>,L2MuonTrajectorySeed> > >"/>
 
-      <class name="L3MuonTrajectorySeed"  ClassVersion="11">
+
+      <class name="L3MuonTrajectorySeed"  ClassVersion="13">
+       <version ClassVersion="13" checksum="3249417346"/>
```
Comment (slava77) [inline DataFormats/MuonSeed/src/classes_def.xml]: please change 13 to 12.
We should not skip versions.

**Comment by slava77** [comment]
review of cloned code is also more complicated: you will get more comments

**Comment by Martin-Grunewald** [comment]
please test

**Comment by slava77** [comment]
please clean up the git history from other PR merges.
The last commit dragged in 30K lines of changes.

**Comment by slava77** [comment]
-1

**Comment by sarafiorendi** [comment]
Yes sorry for confusion, now I cleaned the history up.

**Comment by slava77** [comment]
@cmsbuild please test
[no idea if the extra merges would trigger changes; better test a cleaned up history]

@sarafiorendi thank you

**Comment by fwyzard** [comment]
On 3 March 2016 at 14:33, Slava Krutelyov <notifications@github.com> wrote:

> Is the implementation really so different to justify cloned implementation?
> It may be better to have updates to existing modules.
>

Until the CMS management tells me that we are never going to run over data
that used the legacy GMT data (i.e. up to the end of 2015), we need to have
both versions available.

.Andrea

**Comment by slava77** [comment]
On 3/3/16 6:30 AM, Andrea Bocci wrote:
> On 3 March 2016 at 14:33, Slava Krutelyov <notifications@github.com> wrote:
> 
>> Is the implementation really so different to justify cloned implementation?
>> It may be better to have updates to existing modules.
>>
> 
> Until the CMS management tells me that we are never going to run over data
> that used the legacy GMT data (i.e. up to the end of 2015), we need to have
> both versions available.

Well, that's an extra argument to keep implementation handling legacy
inputs and new inputs  in  one place instead of generating a copy.

**Comment by fwyzard** [comment]
Mhm, no, if the implementation is different enough, there is no reason to
have one module with unnecessarily complex configuration and
implementations, that has to check whether it is using the legacy or
upgraded data for every event.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Martin-Grunewald** [comment]
@slava77 
Could you please have a look and sign?
Thanks!

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+L2MuonSeedGeneratorFromL1T::~L2MuonSeedGeneratorFromL1T(){
+  if (theService)   delete theService;
+  if (theEstimator) delete theEstimator;
+}
+
+void
+L2MuonSeedGeneratorFromL1T::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("GMTReadoutCollection", edm::InputTag("")); // to be removed
+  desc.add<edm::InputTag>("InputObjects",edm::InputTag("hltGmtStage2Digis"));
```
Comment (Martin-Grunewald) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: Why do you change parameters here? Do you want to make my life difficult on purpose?
DO NOT CHANGE ANY PARAMETRS WRT TO THE OLD MODULE VERSION!

**Comment by slava77** [comment]
(I'm not sure what's happening with comments inline to the code)

@Martin-Grunewald , regarding your comment In RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc to
```diff
 +  desc.add<edm::InputTag>("InputObjects",edm::InputTag("hltGmtStage2Digis")); 
```

> Why do you change parameters here? Do you want to make my life difficult on purpose?
> DO NOT CHANGE ANY PARAMETRS WRT TO THE OLD MODULE VERSION!

was the request from trigger to keep the set of old module parameters
(the list of parameter names) or
to also have the values unchanged?
Please clarify.

Given that you signed off, it seemed like the HLT is happy.

**Comment by Martin-Grunewald** [comment]
@slava77 
That was me shouting at the wrong person - deleted.
Yes, TSG is happy with this PR.

**Comment by slava77** [comment]
How do I test this code runtime?

**Comment by Martin-Grunewald** [comment]
@slava77 
This is new code not yet used in any workflow!
It is needed in the release for ConfDB parsing so that HLT can then make use of it.

**Comment by slava77** [comment]
@Martin-Grunewald 
we generally need new code to have a test setup.
At least that's required for all other code in the reco area.

**Comment by Martin-Grunewald** [comment]
Sorry - can't be done - circular dependency.

**Comment by slava77** [comment]
also, there should be an 81X PR.

**Comment by Martin-Grunewald** [comment]
Hmm, we rely on forward port. 

Or perhaps since HLT is the only user, we should move that code
to some HLT package? And move back later when reco wants it so?

**Comment by slava77** [comment]
Similar to other subsystems, I think  RecoMuon/L2MuonSeedGenerator can be removed from reco category.

**Comment by Martin-Grunewald** [comment]
Well, then maybe simply give ownership to HLT and we are fine...

**Comment by Martin-Grunewald** [comment]
... and in the meantime please sign.
Thanks!

**Comment by slava77** [comment]
@davidlange6 can just merge it
not everything has to have reco signature

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
+#ifndef RecoMuon_L2MuonSeedGeneratorFromL1T_L2MuonSeedGeneratorFromL1T_H
+#define RecoMuon_L2MuonSeedGeneratorFromL1T_L2MuonSeedGeneratorFromL1T_H
```
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.h]: the include guard is wrong: it should be ``` RecoMuon_L2MuonSeedGenerator_L2M... ```

**Comment by slava77** [comment]
at least I need to look at the changes.
Maybe OK.
Override if it's urgent.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+  desc.add<string>("Propagator", "");
+  desc.add<double>("L1MinPt",-1.);
+  desc.add<double>("L1MaxEta",5.0);
+  desc.add<unsigned int>("L1MinQuality",0);   
+  desc.addUntracked<bool>("UseOfflineSeed",false);
+  desc.add<bool>("UseUnassociatedL1", true);
+  desc.add<bool>("CentralBxOnly", true);
```
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: this is a new parameter, compared to L2MuonSeedGenerator.cc

```diff
   theSource(iConfig.getParameter<InputTag>("InputObjects")),
-  theL1GMTReadoutCollection(iConfig.getParameter<InputTag>("GMTReadoutCollection")),
+  theL1GMTReadoutCollection(iConfig.getParameter<InputTag>("GMTReadoutCollection")), // to be removed
   thePropagatorName(iConfig.getParameter<string>("Propagator")),
   theL1MinPt(iConfig.getParameter<double>("L1MinPt")),
   theL1MaxEta(iConfig.getParameter<double>("L1MaxEta")),
   theL1MinQuality(iConfig.getParameter<unsigned int>("L1MinQuality")),
   useOfflineSeed(iConfig.getUntrackedParameter<bool>("UseOfflineSeed", false)),
   useUnassociatedL1(iConfig.existsAs<bool>("UseUnassociatedL1") ? 
-                   iConfig.getParameter<bool>("UseUnassociatedL1") : true){
+                   iConfig.getParameter<bool>("UseUnassociatedL1") : true),
+  centralBxOnly_( iConfig.getParameter<bool>("CentralBxOnly") )
+  {
```

I thought the goal was to keep the same set.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+              }
+        
+            }
+          }
+        }
+      }
+  
+    }
+  } else {
+    edm::LogWarning("MissingProduct") << "L1Upgrade muon bx collection not found." << std::endl;
```
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: @Martin-Grunewald 
is this OK?
The old code didn't have a check.

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc (hunk 4 above)
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: about centralBxOnly: IIRC, this was the setting when l1extras were made. Now with the upgrade, it sounds like we let filters decide on their own.
More flexible, but less robust.

Comment (Martin-Grunewald) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: the existAs is not needed as the parameter is in the fillDescriptions.
Parameters added like this are fine...if the default is properly set!

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc (hunk 5 above)
Comment (Martin-Grunewald) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: This is new collection specific

Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: The product should always be present in the event if the path/sequence is appropriately configured.
I find that it's more problematic to let the code to keep going if the only expected product is missing.
hltGmtStage2Digis should always produce the product (even empty).

**Comment by slava77** [comment]
Other than this missing collection effectively exception catching, I think this is good to go for reco.

The include guard fix would be nice, but that's mostly cosmetic.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

File: RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+  edm::Handle<edm::View<TrajectorySeed> > offlineSeedHandle;
+  vector<int> offlineSeedMap;
+  if(useOfflineSeed) {
+    iEvent.getByToken(offlineSeedToken_, offlineSeedHandle);
+    LogTrace(metname) << "Number of offline seeds " << offlineSeedHandle->size() << endl;
+    offlineSeedMap = vector<int>(offlineSeedHandle->size(), 0);
+  }
+
+  if (muColl.isValid()){
```
Comment (slava77) [inline RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc]: The collection not present is more indicative of incorrect configuration.
Why do we need to catch this exception here?
What's a good reason for the MuonBxCollection producer not to make an empty collection?

**Comment by Martin-Grunewald** [comment]
@slava77 
Can you please sign this PR?
Also, should we assign one or more of these packages in the RecoMuon
subsystem to HLT?
* L2MuonIsolationProducer
* L2MuonProducer
* L2MuonSeedGenerator
* L3MuonIsolationProducer
* L3MuonProducer
* L3TrackFinder

**Comment by slava77** [comment]
On 3/7/16 11:27 PM, Martin Grunewald wrote:
> @slava77 <https://github.com/slava77>
> Can you please sign this PR?

I'm waiting for a feedback on my comment.
I think it's wrong to trap the missing collection in
L2MuonSeedGeneratorFromL1T

> Also, should we assign one or more of these packages in the RecoMuon
> subsystem to HLT?
> 
>   * L2MuonIsolationProducer
>   * L2MuonProducer
>   * L2MuonSeedGenerator
>   * L3MuonIsolationProducer
>   * L3MuonProducer
>   * L3TrackFinder
> 

L2MuonSeedGenerator is probably yes.

All other packages relying extensively on reco algorithms.
I'm not sure it's appropriate to pass over maintenance of essentially
reco code to HLT.

**Comment by Martin-Grunewald** [comment]
Fine with me on the packages.
I rather have a LogError than having HLT to crash on a missing product. We can not have that on the filter farm.

**Comment by fwyzard** [comment]
On 8 March 2016 at 14:25, Martin Grunewald <notifications@github.com> wrote:

> I rather have a LogError than having HLT to crash on a missing product. We
> can not have that on the filter farm.


Well, that depends on what can lead to a missing product.

If it is only caused by a mis-configuration, then I'd rather have the crash.

If it is possible to have a missing product (as opposed to an empty
collection) because of the input data, that definitely will not work for
the HLT.

What is the input we are talking about, and what produces it ?

.A

**Comment by Martin-Grunewald** [comment]
Huh, so what if it is a misconfig - do you want to wait for a new patch release, install on the farm,
and restart?

**Comment by fwyzard** [comment]
On 8 March 2016 at 14:38, Martin Grunewald <notifications@github.com> wrote:

> Huh, so what if it is a misconfig - do you want to wait for a new patch
> release, install on the farm,
> and restart?
>
If it is a misconfiguration (meaning a bug in the HLT menu) I expect us to
catch it during the HLT development and validation.
At worst, it can be fixed with an update to the HLT menu.

.A

**Comment by Martin-Grunewald** [comment]
Sorry to say but I find this issue completely besides the point.
Anyway, I guess we get reco signature faster/only if it is taken out, so by all means Sara,
remove this if asap.

**Comment by sarafiorendi** [comment]
Ok, do I remove the check also from 
HLTrigger/Muon/src/HLTMuonL1TRegionalFilter.cc
HLTrigger/Muon/src/HLTL1TMuonSelector.cc
HLTrigger/Muon/src/HLTMuonL1TFilter.cc
?

**Comment by slava77** [comment]
On 3/8/16 6:32 AM, sarafiorendi wrote:
> Ok, do I remove the check also from
> HLTrigger/Muon/src/HLTMuonL1TRegionalFilter.cc
> HLTrigger/Muon/src/HLTL1TMuonSelector.cc
> HLTrigger/Muon/src/HLTMuonL1TFilter.cc
> ?

If these are new checks, they should probably be dropped as well.

If the checks were there before, (IMO, bad choice) they can stay.

Unless LogWarnings are detected very promptly,
there is a danger of inter-path dependencies with the check present.

These are from HLTrigger/Muon
So, Andrea and Martin have the final say.

		--slava

**Comment by fwyzard** [comment]
I'm afraid LogWarning (and even LogError) are going to be completely
ignored in the HLT running online.

We definitely do not want to rely on those to figure out if there are
problems in the data or in the HLT configuration.

(I'm not saying it's a good approach, I'm just being realistic about the
present situation)

**Comment by Martin-Grunewald** [comment]
OK, the BXCollections should always be there (vector size=0 if there is no actual object of that type).
Hence I guess we could remove the check!

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13572	 46b53dd
- code changes are OK, although less code replication is desirable
- jenkins tests pass and comparisons with baseline show no differences

**Comment by davidlange6** [comment]
+1

**Pull request merged**